Code/GraphMol/Descriptors/EEM.cpp

```diff
@@ -0,0 +1,248 @@
+  /* Fill the full n * n block */
+  for (unsigned int i = 0; i < n; i++) {
+    unsigned int t = EEMatoms.EEMatomtype[i];
+    unsigned int idx = EEMatoms.Atomindex[i];
+    double v;
+    if (t == 1) {
+      v = B1[idx];
+    }
+    if (t == 2) {
+      v = B2[idx];
+    }
+    if (t == 3) {
+      v = B3[idx];
+    }
+
+    EEM[i * (n + 1) + i] = v;
+    for (unsigned int j = i + 1; j < n; j++) {
+      EEM[i * (n + 1) + j] = kappa / dist3D[i * n + j];
+      EEM[j * (n + 1) + i] = EEM[i * (n + 1) + j];
+    }
+  }
+  /* Fill last column & row */
+  for (unsigned int i = 0; i < n; i++) {
+    EEM[n * (n + 1) + i] = 1.0;   // column
+    EEM[i * (n + 1) + n] = -1.0;  // row
+  }
+  return std::unique_ptr<double[]>(EEM);
+}
+
+std::unique_ptr<double[]> getBVector(const ROMol &mol, unsigned int n,
+                                     const EEM_arrays &EEMatoms) {
+  /* Fill vector b i.e. -A */
+  double *b = new double[n + 1];
+  for (unsigned int j = 0; j < n; j++) {
+    unsigned int t = EEMatoms.EEMatomtype[j];
+    unsigned int idx = EEMatoms.Atomindex[j];
+    if (t == 1) {
+      b[j] = -A1[idx];
+    }
+    if (t == 2) {
+      b[j] = -A2[idx];
+    }
+    if (t == 3) {
+      b[j] = -A3[idx];
+    }
+  }
+
+  b[n] = MolOps::getFormalCharge(mol);  // sum of charges
+
+  return std::unique_ptr<double[]>(b);
+}
+
+EEM_arrays calculate_EEM_parameters(ROMol mol, unsigned int n) {
+  /* Fill vector b i.e. -A */
+  unsigned int *a = new unsigned int[n];
+  unsigned int *b = new unsigned int[n];
+
+  for (unsigned int j = 0; j < n; j++) {
+    b[j] = getAtomtype(mol, mol.getAtomWithIdx(j));
+    a[j] = mol.getAtomWithIdx(j)->getAtomicNum();
+  }
+
+  return EEM_arrays{a, b};
+}
+
+/* Calculate charges for a particular kappa_data structure */
+void calculate_charges(ROMol mol, double *dist3D, unsigned int numAtoms,
+                       EEM_arrays EEMatoms, std::vector<double> &res) {
+  std::unique_ptr<double[]> A = getEEMMatrix(dist3D, numAtoms, EEMatoms);
+  std::unique_ptr<double[]> b = getBVector(mol, numAtoms, EEMatoms);
+
+  MatrixXd AM = Map<MatrixXd>(A.get(), numAtoms + 1, numAtoms + 1);
+  VectorXd bv = Map<VectorXd>(b.get(), numAtoms + 1);
+  VectorXd Res(numAtoms + 1);
+
+  FullPivLU<MatrixXd> lu(AM);
+  Res = lu.solve(bv);
+
+  for (unsigned int aix = 0; aix < numAtoms; aix++) {
+    res[aix] = Res.data()[aix];
+    mol.getAtomWithIdx(aix)->setProp("_EEMCharge", res[aix], true);
+  }
+}
+
+void getEEMs(const ROMol &mol, std::vector<double> &result,
+             unsigned int numAtoms, int confId) {
+  // 3D distance matrix
+  double *dist3D = MolOps::get3DDistanceMat(mol, confId, false, true);
+  EEM_arrays EEMatoms = calculate_EEM_parameters(mol, numAtoms);
+
+  result.clear();
+  result.resize(numAtoms);
+  calculate_charges(mol, dist3D, numAtoms, EEMatoms, result);
+}
+
+}  // end of anonymous namespace
+
+void EEM(ROMol &mol, std::vector<double> &res, int confId) {
```

Comment:
C++ note, returning std::vector<double> is free as it isn't copied, it is either move constructed or elided.  I think this makes a cleaner interface in general.